feat(db-init): separate database initialization from global database session

mabw-rte opened the pull request:
Context:
- To initialize, communicate with a database, you need to have a session. 
- It is essential to try to the uttermost using the minimum number sessions to avoid alteration conflicts (ideally 1 session/worker) 
- To address this latter point, a global database session is used for all application routes

Issues: 
- In the application initialization step the global session is used to create the database
- The tests are using the same global session in order to  be processed 
- Task-Jobs that were not finished would be tagged to be running after rebooting the application

Solution:
- Use a session for the database initialization
- Task-Jobs that were tagged to be still running or pending are updated to have falied

mabw-rte committed:
feat(db-init): separate database initialization from global database session (#1805)

mabw-rte committed:
feat(db-init): separate database initialization from global database session (#1805)

mabw-rte committed:
feat(db-init): separate database initialization from global database session (#1805)

mabw-rte committed:
feat(db-init): separate database initialization from global database session (#1805)

mabw-rte committed:
feat(db-init): separate database initialization from global database session (#1805)

mabw-rte committed:
feat(db-init): separate database initialization from global database session (#1805)

laurent-laporte-pro commented on antarest/core/tasks/model.py:
```diff
@@ -171,3 +172,17 @@ def __repr__(self) -> str:
         )
+
+
+def cancel_orphan_tasks(engine: Engine, session_args: Mapping[str, bool]) -> None:
```
Pourrais-tu ajouter une docstring ? Il faut expliquer :

> When the web application restarts, such as after a new deployment, any pending or running tasks
> may be lost. To mitigate this, it is preferable to set these tasks to a "FAILED" status. This ensures
> that users can easily identify the tasks that were affected by the restart and take appropriate actions,
> such as restarting the tasks manually.

laurent-laporte-pro commented on antarest/login/main.py:
```diff
@@ -37,7 +37,7 @@ def build_login(
 
     if service is None:
-        user_repo = UserRepository(config)
+        user_repo = UserRepository()
```
Je viens de merger la [PR#1796](https://github.com/AntaresSimulatorTeam/AntaREST/pull/1796) qui contient des `UserRepository()` avec un paramètre `config` dans les tests unitaires. Lors du _merge_ ou du _rebase_, Il faudra faire la modification afin de retirer cette dépendance avec `config`. Je pense que la modification est principalement à faire dans le `conftest.py`.

laurent-laporte-pro commented on antarest/login/model.py:
```diff
@@ -289,3 +302,24 @@ class CredentialsDTO(BaseModel):
     refresh_token: str
+
+
+def init_admin_user(engine: Engine, session_args: Mapping[str, bool], admin_password: str) -> None:
```
Ajouter la docstring. Là aussi, il faut expliquer que l'utilisateur "admin" et le groupe "admin" sont  créés automatiquement lors du démarrage de l'application s'ils n'existent pas. Il convient aussi d'avertir le développeur que cette même fonction peut être appelée dans des processus différents de manière concurrente, car gunicorn démarre plusieurs workers en même temps.

laurent-laporte-pro commented on antarest/main.py:
```diff
@@ -428,6 +437,10 @@ def handle_all_exception(request: Request, exc: Exception) -> Any:
     customize_openapi(application)
+    cancel_orphan_tasks(
+        engine=engine,
+        session_args=SESSION_ARGS,
```
Retirer la virgule à la fin et reformater le code pour que ça rentre sur une seule ligne.

laurent-laporte-pro commented on antarest/main.py:
```diff
@@ -246,7 +249,12 @@ def fastapi_app(
+    application.add_middleware(
+        DBSessionMiddleware,
+        custom_engine=engine,
+        session_args=SESSION_ARGS,
```
Retirer la virgule à la fin et reformater le code pour que ça rentre sur une seule ligne.

laurent-laporte-pro commented on antarest/matrixstore/service.py:
```diff
@@ -54,6 +54,13 @@
+    def __init__(self, matrix_content_repository: MatrixContentRepository) -> None:
+        self.matrix_content_repository = matrix_content_repository
+
+    @property
```
À vérifier, mais je pense que cette propriété n'est pas utiliser…

laurent-laporte-pro commented on antarest/study/main.py:
```diff
@@ -81,6 +81,7 @@ def build_study_service(
     )
 
     generator_matrix_constants = generator_matrix_constants or GeneratorMatrixConstants(matrix_service=matrix_service)
+    generator_matrix_constants.init_constant_matrices()
```
*À VÉRIFIER*

Je pense que si le paramètre `generator_matrix_constants` est `None`, alors on construit une instance de `GeneratorMatrixConstants` et on appelle la méthode `init_constant_matrices()`, mais que si le paramètre n'est pas vide, il est alors inutile appeler la méthode `init_constant_matrices()`.

laurent-laporte-pro commented on antarest/study/storage/variantstudy/business/matrix_constants_generator.py:
```diff
@@ -47,57 +47,66 @@
 ST_STORAGE_INFLOWS = EMPTY_SCENARIO_MATRIX
 
 MATRIX_PROTOCOL_PREFIX = "matrix://"
+MATRIX_CONSTANT_INIT_LOCK_FILE_NAME = "matrix_constant_init.lock"
```
Ne pourrait-on pas trouver un nom de variable plus court, car, au final, cette variable ne devrait être utilisée que dans ce module : on pourrait même préfixer son nom par "_" afin de dire qu'elle est _protégée_.

laurent-laporte-pro commented on tests/core/test_tasks.py:
```diff
@@ -453,3 +465,57 @@ def test_cancel():
+                .filter(TaskJob.status.in_([TaskStatus.RUNNING.value, TaskStatus.PENDING.value]))
+                .all()
+            )
+            assert not updated_task_job
```
Il est préférable de faire une requête qui compte les enregistrements: `COUNT(*)` en SQL.
Le nom de la variable `updated_task_job` n'est pas adéquat.

laurent-laporte-pro commented on tests/core/test_tasks.py:
```diff
@@ -453,3 +465,57 @@ def test_cancel():
+    result_msg: str,
+):
+    max_diff_seconds: int = 1
+    test_id: str = "test_cancel_orphan_tasks_id"
```
utilise un vrai uuid4

laurent-laporte-pro commented on tests/login/test_model.py:
```diff
@@ -1,5 +1,58 @@
+        make_session = sessionmaker(bind=db_engine)
+        with make_session() as session:
+            group = Group(id=GROUP_ID, name=GROUP_NAME)
+            with contextlib.suppress(IntegrityError):
```
Inutile car cette exception ne peut pas être levée.

laurent-laporte-pro commented on tests/login/test_model.py:
```diff
@@ -1,5 +1,58 @@
+        make_session = sessionmaker(bind=db_engine)
+        with make_session() as session:
+            user = User(id=USER_ID, name=USER_NAME, password=Password(TEST_ADMIN_PASS_WORD))
+            with contextlib.suppress(IntegrityError):
```
Inutile car cette exception ne peut pas être levée.

laurent-laporte-pro commented on tests/matrixstore/test_repository.py:
```diff
@@ -165,7 +166,7 @@ def test_datastore_query(self) -> None:
+                    db_session
                     .query(MatrixDataSetRelation)
                     .filter(MatrixDataSetRelation.dataset_id == dataset.id)
                     .all()
```
Utiliser une requête SQL qui compte : `COUNT(*)`.
En fait, on veut vérifier qu'il y a au moins 1 élément.

laurent-laporte-pro commented on tests/study/storage/variantstudy/test_variant_study_service.py:
```diff
@@ -11,6 +11,7 @@
 from antarest.core.requests import RequestParameters
 from antarest.core.utils.fastapi_sqlalchemy import db
 from antarest.login.model import Group, User
+from antarest.matrixstore.repository import MatrixContentRepository
```
*À VÉRIFIER*

Je pense que cet import est superflu.

laurent-laporte-pro commented on tests/variantstudy/model/test_variant_model.py:
```diff
@@ -76,7 +76,7 @@ def test_commands_service(
     ) -> None:
         # Initialize the default matrix constants
         # noinspection PyProtectedMember
```
Supprimer le commentaire

laurent-laporte-pro committed:
test(db-init): update the `user_repo` pytest fixtures which don't require configuration anymore

laurent-laporte-pro committed:
style(db-init): organise imports in `repository.py`

laurent-laporte-pro committed:
test(db-init): change in `TestMatrixRepository` to use the `db_session` fixture

The global object `db` from `fastapi_sqlalchemy` module is no longer used in this unit test

laurent-laporte-pro committed:
refactor(db-init): use `ADMIN_ID` and `ADMIN_NAME` for the administrator ID and name.

laurent-laporte-pro committed:
chore(db-init): remove comments in unit tests

laurent-laporte-pro committed:
test(loging-service): correct and improve the unit tests according the new init-db strategy

mabw-rte committed:
test(formatting): black changes

mabw-rte committed:
feat(pr-review): add doc-strings, formatting, optimize code

laurent-laporte-pro committed:
docs(db-init): improve docstring in database model

laurent-laporte-pro reviewed (approved):
Beware to test on integration server before merging into `dev` ❗

laurent-laporte-pro reviewed (changes requested):
Il y a encore quelques modifications à faire concernant l'intégration de tes changements dans le code legacy :
- Il faut supprimer la méthode `TaskJobService._fix_running_status`.
- Il faut modifier la fonction `build_study_service` car elle ne devrait pas avoir la responsabilité de construire les matrices constantes en appellant `init_constant_matrices`.

Par ailleurs, les services qui tournent en tâche de fond (qui héritent de `IService`, comme le `MatrixGarbageCollector`) devraient utiliser leur propre session de BdD, or ce n'est pas évident car ils utilisent le `StudyService` qui utilise l'instance global `db`. Bref, le sujet n'est pas clos.

laurent-laporte-pro committed:
feat(db-init): remove deprecated method `TaskJobService._fix_running_status`

The `_fix_running_status` method is replaced by `cancel_orphan_tasks`

laurent-laporte-pro committed:
docs(db-model): add `__repr__` in the database model to debug

laurent-laporte-pro committed:
fix(db): correct the session retrieval in `StudyMetadataRepository`

laurent-laporte-pro committed:
feat(db-repo): change the constructor of the `TaskJobRepository` class to accept an optional session parameter

laurent-laporte-pro committed:
chore: correct typo in `exceptions.py`

laurent-laporte-pro committed:
test(fixture): use a `LocalEventBus` for `EventBusService` instead of a mock object

laurent-laporte-pro committed:
perf(service): improve performances by reducing the number of SQL requests in `TaskJobService`

sylvlecl commented on antarest/singleton_services.py:
```diff
@@ -1,90 +1,76 @@
-        for service in self.services_list:
-            self.services_list[service].start(threaded=True)
-
-        self._loop()
```
The removal of this has the following consequence:
the watcher, autoarchiver, etc services are stopped as soon as they are started, they are not kept alive anymore --> to be fixed
